### ISSUE TITLE
Sequel Pro badly trips over columns with backtick (`) in name

### ISSUE THREAD

**Issue by dmoagx** [sequelpro/sequelpro#2105]
A column named `` `strangename` `` is a valid name in MySQL (written as `` ```strangename``` ``) but Sequel Pro will error out on such a table and whats worse, will present an endless loop of warning messages when trying to view the table content.

**Comment by nacho4d** [comment]
I have confirmed that showing a table like the following is problematic in Sequel. As of Sequel 1.0.2, when I select the table from the table list on the left the beach ball appears.

    mysql> CREATE TABLE Persons ( PersonID int, ```LastName``` varchar(255) );
    Query OK, 0 rows affected (0.15 sec)

    mysql> describe Persons
        -> ;
    +------------+--------------+------+-----+---------+-------+
    | Field      | Type         | Null | Key | Default | Extra |
    +------------+--------------+------+-----+---------+-------+
    | PersonID   | int(11)      | YES  |     | NULL    |       |
    | `LastName` | varchar(255) | YES  |     | NULL    |       |
    +------------+--------------+------+-----+---------+-------+
    2 rows in set (0.15 sec)

**Comment by nacho4d** [comment]
I realized that at (least in) OS X 10.10.5 the NSAlert this issue raises was not showing properly. I modified a bit SPOnewayAlertSheet so it accepts NSAlertStyle and reuse that function to show the alert.

https://github.com/sequelpro/sequelpro/pull/2210/files

I will try to look at this issue.

**Comment by nacho4d** [comment]
@dmoagx  I have a question. What should be be output of 

    - (NSUInteger) firstOccurrenceOfCharacter:(unichar)character afterIndex:(NSInteger)startIndex skippingBrackets:(BOOL)skipBrackets ignoringQuotedStrings:(BOOL)ignoreQuotedStrings
    
when `string` is the first line of:

    ```LastName``` varchar(255) DEFAULT NULL
    01234567890123 ← These numbers are just for easy visual counting
    
when `character` is the quote, `startIndex` is -1, `skippingBrackets` is NO (but in this case we don't have brackets so its doesn't mather) and 

①. ignoreQuotedStrings is YES?

Should it be 0? or 2?

②. ignoreQuotedString is NO?

Should it be 11? or 13?

I think these answers will give me hints on what should be fixed. 

<hr/>

I think the reason of this issue is a   

A. missuse of this method. Or  
B. perhaps the method does not fit this case.


If it is B then I would extract the column name

    ``LastName``
    
 with a new method that uses a regex and update the parser accordingly. Then unscape the result to:
 
    `LastName`

So I can finally get the desired value.

**Comment by dmoagx** [comment]
I don't really know much about the parser, but from a quick look the input
```
CREATE TABLE `entities` (↵  `col1` int(11) unsigned) ENGINE=InnoDB
                        ^                          ^              
012345678901234567890123456789012345678901234567890123456789012345
          1         2         3         4         5         6 
```
is handled by the call
```
- firstOccurrenceOfCharacter:')' afterIndex:24 skippingBrackets:YES ignoringQuotedStrings:YES
```
meaning that the function has to handle all
* comments
* stacks of parentheses
* strings (single quote, double quote, backtick quote) and escapes
that occur along the way.

On the other hand this means the function can't be used to find the end of a string by passing quote, doublequote or backtick as `character`.
It doesn't have enough context to actually know it is inside a string and of which type (e.g. double backtick escapes only work inside backtick strings).

The way it currently seems to work is
```
`col1` int(11) unsigned
01234567890123456789012
^    ^                 
          1         2  
```
with the call
```
firstOccurrenceOfCharacter:'`' afterIndex:0 skippingBrackets:NO ignoringQuotedStrings:YES
```

**Comment by rowanbeentje** [comment]
I *think* @nacho4d is looking at the "B" case - the method should return
```
``LastName``
```
...which then gets further de-escaped.  I think the main purpose of this method was to be able to easily grab parts of the SQL, with "ignoring quoted strings" acting as a filter to easily skip quoted column names, comments, etc - and still pick out the rest of the syntax of SQL.  

So I think if you were looking at the string:
```
```LastName(12)``` varchar(255) DEFAULT NULL
01234567890123456789012345678901234567890123
          1         2         3         4
```
...and set is to `ignoreQuotedStrings:YES`, looking for the ` character _should_ pick out position 0 and 17?  And looking for brackets would give 26 and 30?

So basically what @dmoagx said :D

**Comment by nacho4d** [comment]
@dmoagx @rowanbeentje 
> So I think if you were looking at the string:
> 
>     ```LastName(12)``` varchar(255) DEFAULT NULL
>     01234567890123456789012345678901234567890123
>               1         2         3         4
> ...and set is to ignoreQuotedStrings:YES, looking for the ` character should pick out position 0 and 17? > And looking for brackets would give 26 and 30?

Life goal: Get somehow 0 and then 17

So (If I understand well the theory) with  `ignoreQuotedStrings:YES` it means that: **It will skip a quote if the next character is also a quote**.  `ignoreQuotedStrings:NO` means that **it will return the first quote it find, no matter what follows**.

If such logic is correct then it finds character at index 0 is a quote but next one is also a quote so it will keep looking and find character at index 2 is the one to return.
Then, starting form that position, search again and find a quote at index 15 but character at index 16 is also a quote so keep searching until character at index 17.

I got 2 and 17 :(

If above logic is correct, in order to get 0 and 17 ( in [SPSQLParser.m#L503-L510](https://github.com/sequelpro/sequelpro/blob/master/Source/SPSQLParser.m#L503-L510) ) we should find the quotes with

`ignoreQuoteString: NO` and then with `ignoreQuoteString: YES`. 

Is this correct?

**Comment by nacho4d** [comment]
It would be nice it we had unit test cases for these methods to make sure we don't have regressions

**Comment by rowanbeentje** [comment]
I agree, this kind of method is ideal for unit tests, sorry ;)

I think the intention with `ignoreQuotedStrings:YES` was to skip any character which is *inside* a "quoted" string, eg treating the contents as a token. `ignoreQuotedStrings:NO` is just character-by-character matching.

So the intention is to treat:
```
`col1` int(11) unsigned COMMENT "This stuff is 'complex'"
```
as:
```
`$1` int(11) unsigned COMMENT "$2"
```

...whatever the contents.  Returning 2 and 17 is an interesting one :D  Oh boy.  Your suggested fix in SPSQLParser would work, but I think that maybe the `firstOccurrenceOfCharacter:` method should be returning 0 here, recognising that the escaping is happening inside the backticks...?

**Comment by dmoagx** [comment]
Another thing to watch out for:
```
```LastName```
^~~        ~~^
```
At the beginning of the backtick string the characters are interpreted as `backtick + escaped backtick` while at the end they are interpreted as `escaped backtick + backtick`.

So basically we shouldn't be using the `firstOccurrenceOfCharacter:` method here. 
Two alternatives:
* Add a `findEndOfBacktickString:` method and call it in SPSQLParser.m:510 instead when `fromCharacter` and `toCharacter` are both a backtick
* Add a method `extractBacktickQuotedString:` and call that in SPTableData.m:567 instead.
